File: live-server/scripts/check-8002-credit.mjs
/**
 * Rule 8002 (CREDIT-HOUR BAND) — live harness.
 *
 * Drives the Rust 8002 credit-band engine (rule-engine-rs/target/release/check-8002-credit)
 * over the LIVE roster: 8002 computes each crew's MONTHLY credit standalone (FLY →
 * max(4:00, block); ground/off → flat 4:00), prorates the [min,max] band by availability,
 * and warns when outside it. NO 7502 dependency — 8002 fires by itself. Used by the e2e
 * spec `e2e/tests/gantt/rule-8002-credit.spec.ts` (Rule-3xxx).
 *
 * Instance/ruleset resolved from the active RULE workset — do NOT hardcode instance.
 *
 * Data source: every roster_flight activity in the window (FLY block via flight; ground
 * rows credit the flat floor), one TSV row each: crew, group, block_min, YYYY-MM-DD.
 *
 * Usage:
 *   node live-server/scripts/check-8002-credit.mjs [--min 65 --max 75 --from 2026-06-01 --to 2026-07-01] [--persist]
 * Without --persist: prints a JSON summary. With --persist: writes warnings into
 * rule_violation (rule_code 8002 / instance from workset, sev 2, created_by 'rust_8002_credit').
 */
import { spawnSync } from 'node:child_process'
import { createHash } from 'node:crypto'
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import path from 'node:path'
import pg from 'pg'
import { loadRulesetRules, fieldRaw } from './legality-ruleset-params.mjs'

const { Client } = pg
const __dirname = path.dirname(fileURLToPath(import.meta.url))
const REPO = path.resolve(__dirname, '../..')
const ENV_PATH = path.join(REPO, 'live-server', '.env')
const BIN = path.join(REPO, 'rule-engine-rs', 'target', 'release', 'check-8002-credit')

// rule_violation now keys on bigint ruleset_id (= RULE workset, prefer id 103),
// not the dropped varchar rule_group_code column.
const RULE_CODE = '8002'
const SEVERITY = 2 // Overridable / WARNING
const MARKER = 'rust_8002_credit' // created_by — separates these from the BH-window 8002 rows

function hhmmToHours(raw) {
  const s = String(raw ?? '').trim()
  if (s.includes(':')) {
    const [h, m] = s.split(':').map(Number)
    return h + (m || 0) / 60
  }
  return Number(s)
}

/**
 * Credit-band (Type=CH) row from RULE workset 8002 instances, if configured.
 * Current F8 workset 103 may only ship BH/DP — then CLI --min/--max is required.
 */
export async function readRuleMeta(c, opts = {}) {
  const rules = await loadRulesetRules(c, 8002, opts)
  if (!rules.length) throw new Error('No 8002 rule in RULE workset (rule_set)')
  for (const rule of rules) {
    for (const row of rule.rows) {
      if (fieldRaw(row, rule.header, 'Type').toUpperCase() !== 'CH') continue
      const minHours = hhmmToHours(fieldRaw(row, rule.header, 'Min Limits'))
      const maxHours = hhmmToHours(fieldRaw(row, rule.header, 'Max Limits'))
      if (!Number.isFinite(minHours) || !Number.isFinite(maxHours)) {
        throw new Error(`8002/${rule.instance}: invalid CH Min/Max Limits`)
      }
      return {
        rulesetId: rule.rulesetId,
        instance: rule.instance,
        minHours,
        maxHours,
        hasCreditBand: true,
      }
    }
  }
  return {
    rulesetId: rules[0].rulesetId,
    instance: rules[0].instance,
    minHours: null,
    maxHours: null,
    hasCreditBand: false,
  }
}

function readDatabaseUrl() {
  const env = readFileSync(ENV_PATH, 'utf8')
  const line = env.split('\n').find((l) => l.startsWith('DATABASE_URL='))
  if (!line) throw new Error('DATABASE_URL not found in live-server/.env')
  return line.slice('DATABASE_URL='.length).trim().replace(/^["']|["']$/g, '')
}

function argFlag(name, fallback) {
  const i = process.argv.indexOf(name)
  return i >= 0 ? process.argv[i + 1] : fallback
}

// Per-activity credit, DATA-DRIVEN from the assignment fixed credit (same model 7502 uses):
//   FLIGHT (pairing) segment → block × FT (FLT ft=1.0), via pairing_segment → flight.
//   GROUND (no pairing) duty  → fixed_credit_min only.
// Plus a presence row (credit 0) per roster activity so active-days (proration) counts
// every assigned day even when its credit is 0 (days off) or its block is missing.
// Emits one TSV line per row: crew <TAB> credit_minutes <TAB> YYYY-MM-DD.
async function extractActivities(c, from, to) {
  const fly = await c.query(
    `
    SELECT rf.crew_id AS crew_id, COALESCE(f.blk_min, 0)::int AS credit,
           to_char(rf.pday, 'YYYY-MM-DD') AS day
      FROM (SELECT crew_id, pairing_id, MIN(sch_str_dt_utc::date) AS pday
              FROM roster_flight
             WHERE is_deleted = 0 AND assignment_group = 'FLY' AND pairing_id IS NOT NULL
               AND sch_str_dt_utc >= $1::timestamptz AND sch_str_dt_utc < $2::timestamptz
             GROUP BY crew_id, pairing_id) rf
      JOIN pairing_segment ps ON ps.pairing_id = rf.pairing_id AND ps.is_deleted = 0
      JOIN flight f ON f.id = ps.flt_id
    `,
    [from, to],
  )
  const grd = await c.query(
    `
    SELECT rf.crew_id AS crew_id,
           COALESCE(a.fixed_credit_min, 0)::int AS credit,
           to_char(rf.sch_str_dt_utc::date, 'YYYY-MM-DD') AS day
      FROM roster_flight rf
      LEFT JOIN assignment a ON a.assignment = rf.assignment
     WHERE rf.is_deleted = 0 AND rf.assignment_group = 'GRD'
       AND rf.sch_str_dt_utc >= $1::timestamptz AND rf.sch_str_dt_utc < $2::timestamptz
    `,
    [from, to],
  )
  const presence = await c.query(
    `
    SELECT crew_id, to_char(sch_str_dt_utc::date, 'YYYY-MM-DD') AS day
      FROM roster_flight
     WHERE is_deleted = 0
       AND sch_str_dt_utc >= $1::timestamptz AND sch_str_dt_utc < $2::timestamptz
    `,
    [from, to],
  )
  const lines = []
  for (const r of fly.rows) lines.push(`${r.crew_id}\t${r.credit}\t${r.day}`)
  for (const r of grd.rows) lines.push(`${r.crew_id}\t${r.credit}\t${r.day}`)
  for (const r of presence.rows) lines.push(`${r.crew_id}\t0\t${r.day}`)
  return lines.join('\n') + '\n'
}

function runEngine(tsv, minHours, maxHours) {
  const res = spawnSync(
    BIN,
    ['--min-hours', String(minHours), '--max-hours', String(maxHours), '--emit-tsv'],
    { input: tsv, encoding: 'utf8', maxBuffer: 256 * 1024 * 1024 },
  )
  if (res.status !== 0) throw new Error(`check-8002-credit failed: ${res.stderr || res.error}`)
  return res.stdout
    .split('\n')
    .map((l) => l.trim())
    .filter(Boolean)
    .map((l) => {
      const [crewId, month, credit, minPro, maxPro, over, activeDays, daysInMonth] = l.split('\t')
      return {
        crewId,
        month,
        creditMin: Number(credit),
        minProMin: Number(minPro),
        maxProMin: Number(maxPro),
        over: over === '1',
        activeDays: Number(activeDays),
        daysInMonth: Number(daysInMonth),
      }
    })
}

const hhmm = (m) => `${Math.floor(m / 60)}:${String(m % 60).padStart(2, '0')}`

function buildMessage(v) {
  const credit = hhmm(v.creditMin)
  if (v.over) {
    return `Monthly credit ${credit} in 1 CM (${v.month}) exceeds the ${hhmm(v.maxProMin)} maximum (8002 credit band ${hhmm(v.minProMin)}–${hhmm(v.maxProMin)}).`
  }
  return `Monthly credit ${credit} in 1 CM (${v.month}) is below the ${hhmm(v.minProMin)} minimum (8002 credit band ${hhmm(v.minProMin)}–${hhmm(v.maxProMin)}).`
}

async function persist(c, viols, from, to, { rulesetId, ruleInstance }) {
  // Map each crew-month to a triggering June pairing so the gantt (which drops
  // pairing_id NULL) renders it. start_dt = month start (distinct from the BH-window rows).
  const { rows: pairRows } = await c.query(
    `
    SELECT DISTINCT ON (crew_id) crew_id, pairing_id
      FROM roster_flight
     WHERE is_deleted = 0 AND assignment_group = 'FLY' AND pairing_id IS NOT NULL
       AND sch_str_dt_utc >= $1::timestamptz AND sch_str_dt_utc < $2::timestamptz
     ORDER BY crew_id, sch_str_dt_utc
    `,
    [from, to],
  )
  const pairingOf = new Map(pairRows.map((r) => [r.crew_id, Number(r.pairing_id)]))

  await c.query('BEGIN')
  const del = await c.query(`DELETE FROM rule_violation WHERE created_by = $1`, [MARKER])
  let inserted = 0
  const rows = viols.filter((v) => pairingOf.has(v.crewId))
  const BATCH = 500
  for (let off = 0; off < rows.length; off += BATCH) {
    const batch = rows.slice(off, off + BATCH)
    const ph = []
    const vals = []
    batch.forEach((v, idx) => {
      const b = idx * 13
      ph.push(
        `($${b + 1},$${b + 2},$${b + 3},$${b + 4},$${b + 5},$${b + 6},$${b + 7},$${b + 8},$${b + 9},$${b + 10},$${b + 11},$${b + 12},$${b + 13},now(),'${MARKER}','${MARKER}')`,
      )
      const startDt = `${v.month}-01T00:00:00Z`
      const endDt = new Date(Date.UTC(Number(v.month.slice(0, 4)), Number(v.month.slice(5, 7)), 0, 23, 59, 59)).toISOString()
      vals.push(
        v.crewId,
        pairingOf.get(v.crewId),
        rulesetId,
        RULE_CODE,
        ruleInstance,
        startDt,
        endDt,
        SEVERITY,
        Math.round((v.creditMin / 60) * 100) / 100,
        Math.round(((v.over ? v.maxProMin : v.minProMin) / 60) * 100) / 100,
        'HOUR',
        buildMessage(v),
        createHash('sha256').update(`${v.crewId}|${v.month}|8002credit`).digest('hex'),
      )
    })
    await c.query(
      `INSERT INTO rule_violation
         (crew_id, pairing_id, ruleset_id, rule_code, rule_instance,
          start_dt, end_dt, severity, actual_value, limit_value, unit, message, input_hash,
          computed_at, created_by, updated_by)
       VALUES ${ph.join(',')}`,
      vals,
    )
    inserted += batch.length
  }
  await c.query('COMMIT')
  return { deleted: del.rowCount, inserted, skippedNoPairing: viols.length - rows.length }
}

async function main() {
  const from = argFlag('--from', '2026-06-01')
  const to = argFlag('--to', '2026-07-01')
  const doPersist = process.argv.includes('--persist')

  const c = new Client({ connectionString: readDatabaseUrl() })
  await c.connect()
  const ruleMeta = await readRuleMeta(c)
  const minCli = argFlag('--min', null)
  const maxCli = argFlag('--max', null)
  const minHours = minCli != null ? Number(minCli) : ruleMeta.minHours
  const maxHours = maxCli != null ? Number(maxCli) : ruleMeta.maxHours
  if (!Number.isFinite(minHours) || !Number.isFinite(maxHours)) {
    throw new Error(
      '8002 credit band: RULE workset has no Type=CH row — pass --min and --max, or add a CH band to the workset',
    )
  }
  const tsv = await extractActivities(c, from, to)
  const crewMonths = new Set(tsv.split('\n').filter(Boolean).map((l) => l.split('\t')[0])).size
  const viols = runEngine(tsv, minHours, maxHours)
  const over = viols.filter((v) => v.over).length

  let persisted = null
  if (doPersist) persisted = await persist(c, viols, from, to, ruleMeta)
  await c.end()

  const worst = viols.slice().sort((a, b) => b.creditMin - a.creditMin)[0]
  process.stdout.write(
    JSON.stringify({
      rule: `8002/${ruleMeta.instance}`,
      rulesetId: ruleMeta.rulesetId,
      hasCreditBand: ruleMeta.hasCreditBand,
      minHours,
      maxHours,
      crewEvaluated: crewMonths,
      warnings: viols.length,
      overMax: over,
      underMin: viols.length - over,
      worst: worst ? { crewId: worst.crewId, creditHours: Math.round((worst.creditMin / 60) * 10) / 10, over: worst.over } : null,
      ...(persisted ? { persisted } : {}),
    }),
  )
}

if (path.resolve(process.argv[1] ?? '') === fileURLToPath(import.meta.url)) {
  main().catch((err) => {
    process.stderr.write(String(err?.stack ?? err))
    process.exit(1)
  })
}

File: rule-engine-rs/src/bin/check_8002_credit.rs
//! 8002/006 standalone CREDIT-HOUR BAND check (4th row: Type=CH, 1 CM, 75:00/65:00).
//!
//! Reads roster activities on stdin as TSV — one row per activity:
//!     crew_id <TAB> credit_minutes <TAB> YYYY-MM-DD
//! Per-activity credit is computed upstream (node, data-driven from the assignment
//! fixed_credit factors — flight → block×ft; ground → fixed_credit_min only; the same
//! model 7502 uses). This binary sums credit per crew per
//! CALENDAR MONTH, prorates the band by availability (active days / days in month), and
//! warns when the monthly credit is above Max or below Min.
//!
//! Usage:
//!     check-8002-credit [--min-hours 65] [--max-hours 75] [--top 15] [--emit-tsv]
//!
//! --emit-tsv prints one violation per line for piping into rule_violation:
//!     crew <TAB> month <TAB> credit_min <TAB> min_pro <TAB> max_pro <TAB> over(1|0)
//!          <TAB> active_days <TAB> days_in_month

use std::collections::BTreeMap;
use std::io::{self, Read};

use rois_rule_engine::{
    format_hhmm,
    rules::rule8002::{check_credit_band, days_in_month, CreditBandViolation},
};

fn arg_value(args: &[String], flag: &str) -> Option<String> {
    args.iter()
        .position(|a| a == flag)
        .and_then(|i| args.get(i + 1).cloned())
}

#[derive(Default)]
struct MonthAcc {
    credit: i64,
    active_days: std::collections::BTreeSet<i64>, // day-of-month (or ordinal) for availability
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let min_hours: f64 = arg_value(&args, "--min-hours")
        .and_then(|v| v.parse().ok())
        .unwrap_or(65.0);
    let max_hours: f64 = arg_value(&args, "--max-hours")
        .and_then(|v| v.parse().ok())
        .unwrap_or(75.0);
    let top: usize = arg_value(&args, "--top")
        .and_then(|v| v.parse().ok())
        .unwrap_or(15);
    let raw_min = (min_hours * 60.0).round() as i64;
    let raw_max = (max_hours * 60.0).round() as i64;

    let mut input = String::new();
    io::stdin().read_to_string(&mut input).expect("read stdin");

    // (crew, "YYYY-MM") -> accumulator
    let mut acc: BTreeMap<(String, String), MonthAcc> = BTreeMap::new();
    let mut parsed = 0usize;
    let mut skipped = 0usize;
    for line in input.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        // TSV: crew <TAB> credit_minutes <TAB> YYYY-MM-DD. Credit is pre-computed upstream
        // (node, data-driven from the assignment fixed_credit_min factor — same
        // model 7502 uses). One row per roster activity (credit 0 marks an active day only).
        let mut it = line.split('\t');
        let (crew, credit, day) = match (it.next(), it.next(), it.next()) {
            (Some(c), Some(cr), Some(d)) => (c, cr, d),
            _ => {
                skipped += 1;
                continue;
            }
        };
        let credit_min: i64 = match credit.parse() {
            Ok(v) => v,
            Err(_) => {
                skipped += 1;
                continue;
            }
        };
        if day.len() < 10 {
            skipped += 1;
            continue;
        }
        let month = day[0..7].to_string(); // YYYY-MM
        let dom: i64 = day[8..10].parse().unwrap_or(0);
        let e = acc.entry((crew.to_string(), month)).or_default();
        e.credit += credit_min;
        e.active_days.insert(dom);
        parsed += 1;
    }

    let t0 = std::time::Instant::now();
    let mut viols: Vec<CreditBandViolation> = Vec::new();
    let mut active_days_of: Vec<(String, i64, i64)> = Vec::new(); // crew, active, dim — for report
    for ((crew, month), a) in &acc {
        let y: i64 = month[0..4].parse().unwrap_or(2026);
        let m: i64 = month[5..7].parse().unwrap_or(1);
        let dim = days_in_month(y, m);
        let factor = (a.active_days.len() as f64 / dim as f64).min(1.0);
        active_days_of.push((crew.clone(), a.active_days.len() as i64, dim));
        if let Some(v) = check_credit_band(crew, month, a.credit, raw_min, raw_max, factor) {
            viols.push(v);
        }
    }
    let eval = t0.elapsed();
    let total_crew_months = acc.len();
    viols.sort_by(|a, b| b.credit_minutes.cmp(&a.credit_minutes));

    if args.iter().any(|a| a == "--emit-tsv") {
        let dim_map: std::collections::HashMap<String, (i64, i64)> = active_days_of
            .iter()
            .map(|(c, ad, dm)| (c.clone(), (*ad, *dm)))
            .collect();
        for v in &viols {
            let (ad, dm) = dim_map.get(&v.crew_id).copied().unwrap_or((0, 30));
            println!(
                "{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
                v.crew_id,
                v.month,
                v.credit_minutes,
                v.min_minutes,
                v.max_minutes,
                if v.over { 1 } else { 0 },
                ad,
                dm,
            );
        }
        eprintln!(
            "emitted {} credit-band warnings over {} crew-months in {:.3} ms",
            viols.len(),
            total_crew_months,
            eval.as_secs_f64() * 1000.0,
        );
        return;
    }

    let over = viols.iter().filter(|v| v.over).count();
    let under = viols.len() - over;
    println!("─────────────────────────────────────────────────────────────");
    println!(" Rule 8002/006 — CREDIT-HOUR BAND (Type=CH, 1 CM) — Rust engine");
    println!(
        " band: credit must be within {:.0}:00 … {:.0}:00 per calendar month (prorated)",
        min_hours, max_hours
    );
    println!("─────────────────────────────────────────────────────────────");
    println!(
        " activities    : {}{}",
        parsed,
        if skipped > 0 {
            format!("  ({} skipped)", skipped)
        } else {
            String::new()
        }
    );
    println!(" crew-months    : {}", total_crew_months);
    println!(" eval time      : {:.3} ms", eval.as_secs_f64() * 1000.0);
    println!(
        " WARNINGS       : {}  ({} over max, {} under min)",
        viols.len(),
        over,
        under
    );
    println!("─────────────────────────────────────────────────────────────");
    println!(" Top {} by monthly credit:", top.min(viols.len()));
    println!(
        "   {:<12} {:<8} {:>9} {:>9} {:>9}   {}",
        "crew", "month", "credit", "min", "max", "bound"
    );
    for v in viols.iter().take(top) {
        println!(
            "   {:<12} {:<8} {:>9} {:>9} {:>9}   {}",
            v.crew_id,
            v.month,
            format_hhmm(v.credit_minutes),
            format_hhmm(v.min_minutes),
            format_hhmm(v.max_minutes),
            if v.over { "OVER max" } else { "under min" }
        );
    }
    println!("─────────────────────────────────────────────────────────────");
}
